File: Python/Birthday_calc.py
# Birthday_calc.py
from datetime import date


def main() -> None:
    raw = input("Enter your birthday (YYYY-MM-DD): ").strip()
    try:
        bday = date.fromisoformat(raw)
    except ValueError:
        print("Invalid date format. Use YYYY-MM-DD.")
        return

    today = date.today()
    if bday > today:
        print("Birthday is in the future.")
        return

    # Age
    age = today.year - bday.year - ((today.month, today.day) < (bday.month, bday.day))

    # Next birthday
    try:
        next_bd = bday.replace(year=today.year)
    except ValueError:  # Feb 29 in non-leap year
        next_bd = date(today.year, 3, 1)
    if next_bd < today:
        try:
            next_bd = bday.replace(year=today.year + 1)
        except ValueError:
            next_bd = date(today.year + 1, 3, 1)

    days_lived = (today - bday).days
    days_until = (next_bd - today).days
    weekday = bday.strftime("%A")

    print(f"You were born on a {weekday}.")
    print(f"Age: {age} years")
    print(f"Days lived: {days_lived:,}")
    if days_until == 0:
        print("🎉 Happy birthday — it is today!")
    else:
        print(
            f"Days until next birthday: {days_until} ({next_bd.strftime('%A, %B %d, %Y')})"
        )


if __name__ == "__main__":
    main()

File: Typescript/Birthday_calc.ts
const readline = require('node:readline/promises');
const { stdin, stdout } = require('node:process');

const isLeapYear = (year) =>
  year % 4 === 0 && (year % 100 !== 0 || year % 400 === 0);

const parseDate = (text) => {
  const match = /^(\d{4})-(\d{2})-(\d{2})$/.exec(text.trim());
  if (!match) {
    return null;
  }

  const year = Number(match[1]);
  const month = Number(match[2]);
  const day = Number(match[3]);
  const date = new Date(year, month - 1, day);

  if (
    date.getFullYear() !== year ||
    date.getMonth() !== month - 1 ||
    date.getDate() !== day
  ) {
    return null;
  }

  return date;
};

const dateOnly = (date) =>
  new Date(date.getFullYear(), date.getMonth(), date.getDate());

const dayNames = [
  'Sunday',
  'Monday',
  'Tuesday',
  'Wednesday',
  'Thursday',
  'Friday',
  'Saturday',
];

const monthNames = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December',
];

const formatLongDate = (date) =>
  `${dayNames[date.getDay()]}, ${monthNames[date.getMonth()]} ${String(
    date.getDate(),
  ).padStart(2, '0')}, ${date.getFullYear()}`;

const main = async () => {
  const rl = readline.createInterface({ input: stdin, output: stdout });

  try {
    const raw = await rl.question('Enter your birthday (YYYY-MM-DD): ');
    const bday = parseDate(raw);

    if (!bday) {
      console.log('Invalid date format. Use YYYY-MM-DD.');
      return;
    }

    const today = dateOnly(new Date());
    const birthday = dateOnly(bday);

    if (birthday > today) {
      console.log('Birthday is in the future.');
      return;
    }

    let age = today.getFullYear() - birthday.getFullYear();
    const hasHadBirthday =
      today.getMonth() > birthday.getMonth() ||
      (today.getMonth() === birthday.getMonth() &&
        today.getDate() >= birthday.getDate());
    if (!hasHadBirthday) {
      age -= 1;
    }

    const buildBirthday = (year) => {
      if (birthday.getMonth() === 1 && birthday.getDate() === 29 && !isLeapYear(year)) {
        return new Date(year, 2, 1);
      }
      return new Date(year, birthday.getMonth(), birthday.getDate());
    };

    let nextBirthday = buildBirthday(today.getFullYear());
    if (nextBirthday < today) {
      nextBirthday = buildBirthday(today.getFullYear() + 1);
    }

    const dayMs = 24 * 60 * 60 * 1000;
    const daysLived = Math.floor((today - birthday) / dayMs);
    const daysUntil = Math.floor((nextBirthday - today) / dayMs);

    console.log(`You were born on a ${dayNames[birthday.getDay()]}.`);
    console.log(`Age: ${age} years`);
    console.log(`Days lived: ${daysLived.toLocaleString()}`);

    if (daysUntil === 0) {
      console.log('Happy birthday - it is today!');
    } else {
      console.log(
        `Days until next birthday: ${daysUntil} (${formatLongDate(nextBirthday)})`,
      );
    }
  } finally {
    rl.close();
  }
};

main().catch((error) => {
  console.error('Unexpected error:', error);
});
